es6-shim.js

```diff
@@ -756,6 +771,64 @@
   });
   addIterator(ArrayIterator.prototype);
 
+  var ObjectIterator = function (object, kind) {
+    this.object = object;
+    this.i = 0;
+    // Don't generate keys yet.
+    this.array = null;
+    this.kind = kind;
+  };
+
+  defineProperties(ObjectIterator.prototype, {
+    next: function () {
+      var key, array = this.array;
+      if (!(this instanceof ObjectIterator)) {
+        throw new TypeError('Not an ObjectIterator');
+      }
+
+      if (typeof array !== 'undefined') {
```

Comment:
I wonder if it would be better, instead of keeping an array of all keys and iterating over it, to continually `unshift` it in a loop? That way, when it's not an array, you need to populate it, and when it's an empty array, you know it's done.

That would eliminate the need for a state enum as well as 1 of the 3 states, and it would also eliminate the need for `this.i`.